safetensors/src/slice.rs

```diff
@@ -0,0 +1,418 @@
+use crate::TensorView;
+use std::ops::{
+    Bound, Range, RangeBounds, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive,
+};
+
+/// Error representing invalid slicing attempt
+#[derive(Debug)]
+pub enum InvalidSlice {
+    TooManySlices,
+}
+
+#[derive(Debug)]
+pub enum TensorIndexer {
+    //Select(usize),
+    Narrow(Bound<usize>, Bound<usize>),
+    //IndexSelect(Tensor),
+}
+
+// impl From<usize> for TensorIndexer {
+//     fn from(index: usize) -> Self {
+//         TensorIndexer::Select(index)
+//     }
+// }
+
+// impl From<&[usize]> for TensorIndexer {
+//     fn from(index: &[usize]) -> Self {
+//         let tensor = index.into();
+//         TensorIndexer::IndexSelect(tensor)
+//     }
+// }
+//
+// impl From<Vec<usize>> for TensorIndexer {
+//     fn from(index: Vec<usize>) -> Self {
+//         let tensor = Tensor::of_slice(&index);
+//         TensorIndexer::IndexSelect(tensor)
+//     }
+// }
+
+macro_rules! impl_from_range {
+    ($range_type:ty) => {
+        impl From<$range_type> for TensorIndexer {
+            fn from(range: $range_type) -> Self {
+                use std::ops::Bound::*;
+
+                let start = match range.start_bound() {
+                    Included(idx) => Included(*idx),
+                    Excluded(idx) => Excluded(*idx),
+                    Unbounded => Unbounded,
+                };
+
+                let end = match range.end_bound() {
+                    Included(idx) => Included(*idx),
+                    Excluded(idx) => Excluded(*idx),
+                    Unbounded => Unbounded,
+                };
+
+                TensorIndexer::Narrow(start, end)
+            }
+        }
+    };
+}
+
+impl_from_range!(Range<usize>);
+impl_from_range!(RangeFrom<usize>);
+impl_from_range!(RangeFull);
+impl_from_range!(RangeInclusive<usize>);
+impl_from_range!(RangeTo<usize>);
+impl_from_range!(RangeToInclusive<usize>);
+
+pub trait IndexOp<'data, T> {
```

Comment:
I'm not seeing where this trait is used for now.
Really not a big fan of the super short name `i` for its main method.

Reply:
The trait is here only to support "multiple signature" functions (which isn't allowed).

This works by creating a trait and 1 function that works on anyone that has that trait, and then implement the trait for all types of arguments we want to handle.

I really copied this: https://docs.rs/tch/latest/tch/index/index.html which aims to make something as easy to use as pytorch.
And we need some sort of trait because `Range`, `RangeFull` at least are different types, so we cannot have the sliced data accept a concrete type directly.

The "no trait" route is doing an enum:
```rust
enum TensorIndexer{
    Full(RangeFull)
    Range(Range<usize>)
    ...
}
```
which does work but it much more tedious on the caller.

As for the name of the function, we definitely can make it more explicit here. It's intentionally small in `tch-rs` because it's meant to be used directly and mimic Python indexing (I can dig it up, but using real slicing in Rust, has some kind of limitations because of the trait it asks to implement).

Here it's the same. For Python users, they should never see it, but for rust users doing something like

```rust
let tensorview =  SafeTensors::deserialize(&content).unwrap().tensor("mytensor");
let buffer: Vec<u8> = tensorview.i( (.., ..2,3..4) ).unwrap()..flat_map(|b| b.to_vec()).collect();
```

Feels better than the non trait version:
```rust
let tensorview =  SafeTensors::deserialize(&content).unwrap().tensor("mytensor");
let buffer: Vec<u8> = tensorview.get_sliced_data( vec![
  TensorIndexer::Narrow(Bound::Unbounded, Bound::Unbounded),  
  TensorIndexer::Narrow(Bound::Unbounded, Bound::Excluded(2)),
  TensorIndexer::Narrow(Bound::Included(3), Bound::Excluded(4)),
).unwrap()..flat_map(|b| b.to_vec()).collect();
```

So purely ergonomics. Naming it `slice` (or some other name) is really doable. I kind of enjoy `i` in `tch-rs` though.

Reply:
Thanks for the explanation! Would prefer slice to `i` personally (anything more descriptive than `i` ;-) ) but that's my personal taste.

Reply:
Moved to `slice`. No hugely strong opinions yet.